bridge-ffi/src/lib.rs

```diff
@@ -0,0 +1,650 @@
+#![allow(non_camel_case_types)]
+
+use prost::Message;
+use std::str::FromStr;
+use temporal_sdk_core_protos::coresdk::bridge;
+
+extern crate libc;
+
+/// A set of bytes owned by Core. This must always be passed to
+/// tmprl_bytes_free when no longer in use.
+#[repr(C)]
+pub struct tmprl_bytes_t {
+    bytes: *const u8,
+    len: libc::size_t,
+    cap: libc::size_t,
+    disable_free: bool,
+}
+
+impl tmprl_bytes_t {
+    fn from_vec(vec: Vec<u8>) -> tmprl_bytes_t {
+        // Mimics Vec::into_raw_parts that's only available in nightly
+        let mut vec = std::mem::ManuallyDrop::new(vec);
+        tmprl_bytes_t {
+            bytes: vec.as_mut_ptr(),
+            len: vec.len(),
+            cap: vec.capacity(),
+            disable_free: false,
+        }
+    }
+
+    fn from_vec_disable_free(vec: Vec<u8>) -> tmprl_bytes_t {
+        let mut b = tmprl_bytes_t::from_vec(vec);
+        b.disable_free = true;
+        b
+    }
+
+    fn into_raw(self) -> *mut tmprl_bytes_t {
+        Box::into_raw(Box::new(self))
+    }
+}
+
+// We only impl this because they are required by lazy_static and raw pointers
+// are not usually safe for send
+unsafe impl Send for tmprl_bytes_t {}
+unsafe impl Sync for tmprl_bytes_t {}
```

Comment:
Do we need to even use lazy static for these defaults? Could just construct them every time and avoid this. If we do this, we have to ensure all C side implementations never use thread-local locking etc, or they could end up violating safety

Reply:
Yeah, it may be an over-optimization.

> If we do this, we have to ensure all C side implementations never use thread-local locking etc, or they could end up violating safety

This shouldn't be affected by any of that because from a caller/lang perspective, `tmprl_bytes_t` are immutable.

Reply:
Ah, OK good point. We should explicitly mention that in the docstring too, C/lang should never modify the bytes.

Reply:
> We should explicitly mention that in the docstring too, C/lang should never modify the bytes.

Ha, just did this pending push anyways (also mentioned `cap` and `disable_free` are for internal use only)

Reply:
Done